crates/runtime/src/instance/allocator/pooling/uffd.rs

```diff
@@ -0,0 +1,833 @@
+//! Implements user space page fault handling with the `userfaultfd` ("uffd") system call on Linux.
+//!
+//! Handling page faults for memory accesses in regions relating to WebAssembly instances
+//! enables the implementation of protecting guard pages in user space rather than kernel space.
+//!
+//! This reduces the number of system calls and kernel locks needed to provide correct
+//! WebAssembly memory semantics.
+//!
+//! Additionally, linear memories can be lazy-initialized upon access.
+//!
+//! This feature requires a Linux kernel 4.11 or newer to use.
+
+use super::{InstancePool, StackPool};
+use crate::{instance::Instance, Mmap};
+use std::convert::TryInto;
+use std::ptr;
+use std::sync::{
+    atomic::{AtomicBool, Ordering},
+    Arc,
+};
+use std::thread;
+use userfaultfd::{Event, FeatureFlags, IoctlFlags, Uffd, UffdBuilder};
+use wasmtime_environ::{entity::EntityRef, wasm::DefinedMemoryIndex, MemoryInitialization};
+
+const WASM_PAGE_SIZE: usize = wasmtime_environ::WASM_PAGE_SIZE as usize;
+
+pub unsafe fn make_accessible(_addr: *mut u8, _len: usize) -> bool {
+    // A no-op when userfaultfd is used
+    true
+}
+
+pub unsafe fn reset_guard_page(addr: *mut u8, len: usize) -> bool {
+    // Guard pages are READ_WRITE with uffd until faulted
+    region::protect(addr, len, region::Protection::READ_WRITE).is_ok()
+}
+
+pub unsafe fn decommit(addr: *mut u8, len: usize) {
+    // Use MADV_DONTNEED to mark the pages as missing
+    // This will cause a missing page fault for next access on any page in the given range
+    assert_eq!(
+        libc::madvise(addr as _, len, libc::MADV_DONTNEED),
+        0,
+        "madvise failed to mark pages as missing: {}",
+        std::io::Error::last_os_error()
+    );
+}
+
+pub fn create_memory_map(_accessible_size: usize, mapping_size: usize) -> Result<Mmap, String> {
+    // Allocate a single read-write region at once
+    // As writable pages need to count towards commit charge, use MAP_NORESERVE to override.
+    // This implies that the kernel is configured to allow overcommit or else
+    // this allocation will almost certainly fail without a plethora of physical memory to back the allocation.
+    // The consequence of not reserving is that our process may segfault on any write to a memory
+    // page that cannot be backed (i.e. out of memory conditions).
+
+    if mapping_size == 0 {
+        return Ok(Mmap::new());
+    }
+
+    unsafe {
+        let ptr = libc::mmap(
+            ptr::null_mut(),
+            mapping_size,
+            libc::PROT_READ | libc::PROT_WRITE,
+            libc::MAP_PRIVATE | libc::MAP_ANON | libc::MAP_NORESERVE,
```

Comment:
Here's my understanding of [over-committing memory on Linux](https://www.kernel.org/doc/Documentation/vm/overcommit-accounting):

A lot of distros use the default of `0` for `vm.overcommit_memory`.  This causes the kernel to apply a heuristic to see if it can likely satisfy a request to physically back all the requested accessible pages in the future.  If the kernel thinks it won't be able to physically back the allocation, it fails.

`MAP_NORESERVE` effectively tells a kernel using this over-commit strategy to forgo this check.  In this case, we're pretty much guaranteed that the kernel will not be able to physically back such a gigantic VM allocation.  However, only a fraction of the pages will ever need backing, so `MAP_NORESERVE` is our way of telling the kernel we know better than it does.

I believe that a setting of `1` for `vm.overcommit_memory` is the same as every allocation being `MAP_NORESERVE`.

A setting of `2` will always cause the pooling allocator to fail its allocation.  In this case, `MAP_NORESERVE` has no meaning.

Ultimately this means we're at risk of random `SIGSEGV` for accesses to otherwise accessible pages in OOM conditions because we told the kernel to give us no guarantees that the page can be backed (I believe it might also sidestep the OOM killer for such pages, so the kernel won't bother killing other processes to give you backing space).

Reply:
So if you set `vm.overcommit_memory` to 2, wasmtime with uffd (or I guess the pooling allocator in general?) is inoperable.

Otherwise this business about a random `SIGSEGV`, how does that play out with uffd? We're the ones who ask  the kernel "please allocate a zero page for that spot" or copy from somewhere else. Does that request for a page fail with an error we can handle? Or does it "succeed" and/or kill our process right there?

Reply:
Random faults that would otherwise signal a `SIGSEGV` aren't sent to our fault handler as we only monitor specific regions of address space and they are protected in such a way (i.e .`read+write`) that the kernel should never signal a `SIGSEGV` normally itself (other than, perhaps as a failure to back one of our pages given we're massively over-committing the address space reservation).

Reply:
So I'm going to push a change that removes the use of `MAP_NORESERVE`.  This was previously needed because the memory pool was being protected as `READ_WRITE` for the entire mapping (which is multiple TiB by default for 64-bit hosts).

We now map the entire memory pool with protection `NONE`, which doesn't count towards overcommit limits.

When `uffd` is enabled, it marks *only* the pages that could potentially be accessible as `READ_WRITE` (and thus only this does count towards the overcommit limits).  This also means the user fault handler now *only* gets faults for pages that might be accessible depending on the linear memory's current size.

This has the benefit of failing fast (when the pooling strategy is set in the `Config`) if the kernel doesn't think it can reserve *all* the backing space needed for the accessible pages; compared to receiving random `SIGSEGV` with `MAP_NORESERVE`, I think this is desirable.

Resolving this conversation for now.